Fix react/http-client 0.4.13 response body issue

shulard opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | yes |
| Deprecations? | no |
| Related tickets | fixes #17 |
| Documentation |  |
| License | MIT |
#### What's in this PR?

This PR has been pushed to fixed the compatibility error with `react/http-client` v0.4.13.
#### Why?

There was a break change published in v0.4.13 around response body type. Before the first response body was sent as a `StreamInterface` instance, now it's "just" string.

Before we are using that `StreamInterface` instance as `$bodyStream`... Now it has been updated to use the `Http\Message\StreamFactory` to build a valid StreamInterface instance.

If a BC Break because I've introduced a new constructor parameter to be able to inject `StreamFactory` instance...

Of course the `Http\Discovery\StreamFactoryDiscovery` object is used when parameter is empty.

With this change, I've tested on lowest dependencies and it works too.
#### Checklist
- [ ] Updated CHANGELOG.md to describe BC breaks / deprecations | new feature | bugfix
#### To Do
- [x] Validate PR approach...
- [ ] Create a release with a composer update to disallow react/http-client > 0.4.12 ?

joelwurtz commented on src/Client.php:
```diff
@@ -44,14 +46,21 @@ class Client implements HttpClient, HttpAsyncClient
      */
     public function __construct(
         ResponseFactory $responseFactory = null,
+        StreamFactory $streamFactory = null,
```
This is BC Break, you should add it to the last parameter of the constructor, it would still be BC break but less given people may use the discovery system.

joelwurtz commented:
If we use the StreamFactory as the last parameter of the constructor and we add php-http/message to the dependencies tree there will be no BC break, as react/http-client already include guzzle/psr7 interface and this library has a dependency on php-http/discovery then you will always have a StreamFactory service even if it was not injected.

shulard commented:
Ok 😅 so I'll update my code!

joelwurtz commented:
LGTM,

@dbu @sagikazarmark Ok for you ? IMO there is no bc break here.

sagikazarmark commented:
As long as the comment above is resolved OK for me

joelwurtz commented on src/Client.php:
```diff
@@ -94,17 +104,16 @@ public function sendAsyncRequest(RequestInterface $request)
             $reactResponse->on('data', function ($data) use (&$bodyStream) {
                 if ($data instanceof StreamInterface) {
-                    $bodyStream = $data;
+                    $bodyStream->write($data->getContents());
```
> Maybe I can use the incoming stream as $bodyStream like before ?

I think this will be very hard :/. There is no way to avoid memory problem here as react want us to work with data and not stream so the react interface block us from doing anything.

The only way to do that is to be sure that we create a stream using `php://temp` by example (not in memory, or at least not to much) and then read -> write by chunk of specified size (8192 is the best as it's the default chunk for PHP so there is no internal buffer).

But i don't think we should do this as the last version of react only give us data as a string and not a StreamInterface so this would be an enhancement for old version (so when people will update this lib with the enhancement they will certainly also update react/http-client which does not provide a StreamInterface anymore, so this enhancement will never be used...).

> Not to mention getContents reads the remainder, not the whole stream

:+1: We should only cast the stream to string `$bodyStream->write((string) $data);` in the same way that react as update it's library.

joelwurtz replied:
In fact you could remove the condition and only do : 

```
$bodyStream->write((string) $data);
```

shulard commented:
I've updated the code with string cast, much simpler 😄.

Nyholm commented on composer.json:
```diff
@@ -16,7 +16,8 @@
         "react/stream": "^0.4.3",
-        "php-http/discovery": "^1.0"
+        "php-http/discovery": "^1.0",
+        "php-http/message": "^1.0"
```
Why is php-http/message added?

joelwurtz replied:
This library now require a StreamFactory to work which is added in the constructor, but it can be set to null and you the discovery system to work.

So if we keep it like that, this will be BC for people setting all the contrusctor parameters in order to avoid using Discovery, as the new parameter cannnot be found.

However `react/http-client` depends on ̀guzzlehttp/psr7` so by adding `php-http/message` we are sure to found a service for the StreamFactory in all cases so there is no BC.

Nyholm replied:
Okey, I've missed that discussion. 
I agree. But I would like to remove that line before tagging 1.0 (and preferably before tagging 0.3.0).

Nyholm commented on src/Client.php:
```diff
@@ -94,17 +104,12 @@ public function sendAsyncRequest(RequestInterface $request)
-                } else {
-                    $bodyStream->write($data);
-                }
+                $bodyStream->write((string)$data);
```
Code style: `(string) $data`

shulard replied:
Updated

shulard commented:
Hello !

To address #20 in the same time (and avoid adding / moving the line in two different commits) I've already moved `php-http/message` dependency to `require-dev` section.

I also added a changelog update. If everyone is OK, I'll merge the PR asap.
